### ISSUE TITLE
Unexpected spatial_soft_argmax2d output

### ISSUE THREAD

**Issue by markdjwilliams** [kornia/kornia#2498]
### Describe the bug

I have a heatmap image (below) where the location of the blob's centroid is at approximately [18, 31], somewhat left of center. 

However, when I pass this image to `spatial_soft_argmax2d()` it reports the centroid to be at [31, 31] - almost at the midpoint of the 64x64 heatmap.

![heatmap](https://github.com/kornia/kornia/assets/25598354/5009aeb0-2bb2-4fab-9b44-fa5c2ae0f5c3)

### Reproduction steps

The following script reconstructs the image and queries the heatmap's centroid:

```python
import torch
from kornia.geometry.subpix import spatial_soft_argmax2d   

x = torch.tensor([26, 27, 27, 28, 28, 28, 28, 28, 28, 28, 28, 28, 29, 29, 29, 29, 29,
       29, 29, 29, 29, 29, 29, 29, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30,
       30, 30, 30, 30, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31,
       32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 33, 33, 33, 33,
       33, 33, 33, 33, 33, 33, 33, 33, 33, 34, 34, 34, 34, 34, 34, 34, 34,
       34, 34, 34, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 36, 36, 36,
       36, 36, 36, 36, 36, 36, 36, 36, 37, 37, 37, 37, 37, 37, 37, 37, 37,
       37, 37, 38, 38, 38, 38, 38, 38, 38, 38, 39, 39, 39, 39, 39, 48, 54,
       54, 54, 56, 62, 62, 62])
y = torch.tensor([40, 17, 21, 15, 16, 17, 18, 19, 20, 21, 22, 23, 14, 15, 16, 17, 18,
       19, 20, 21, 22, 23, 24, 41, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
       23, 24, 25, 41, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 28,
       13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 37, 38, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 37, 38, 13, 14, 15, 16, 17, 18, 19, 20,
       21, 22, 23, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 13, 14, 15,
       16, 17, 18, 19, 20, 21, 22, 23, 13, 14, 15, 16, 17, 18, 19, 20, 21,
       22, 23, 15, 16, 17, 18, 19, 20, 21, 22, 16, 17, 18, 19, 20, 42, 28,
       29, 30, 28,  1,  2,  3])
v = torch.tensor([0.00392157, 0.00392157, 0.00392157, 0.01176471, 0.01568628,
       0.01568628, 0.01568628, 0.01176471, 0.01960784, 0.02352941,
       0.01568628, 0.01568628, 0.01960784, 0.05490196, 0.09019608,
       0.10196079, 0.09803922, 0.07843138, 0.05882353, 0.03921569,
       0.02352941, 0.02352941, 0.00784314, 0.00392157, 0.00392157,
       0.05882353, 0.14117648, 0.22745098, 0.27450982, 0.27450982,
       0.21960784, 0.14117648, 0.07843138, 0.03137255, 0.02352941,
       0.00784314, 0.00392157, 0.00392157, 0.01568628, 0.10588235,
       0.2509804 , 0.4       , 0.49411765, 0.5058824 , 0.4117647 ,
       0.27058825, 0.14117648, 0.04313726, 0.00784314, 0.00392157,
       0.00392157, 0.03137255, 0.14509805, 0.3372549 , 0.5372549 ,
       0.6784314 , 0.7019608 , 0.58431375, 0.39607844, 0.20784314,
       0.0627451 , 0.00784314, 0.00392157, 0.00392157, 0.03529412,
       0.15686275, 0.36078432, 0.5803922 , 0.74509805, 0.78039217,
       0.654902  , 0.4509804 , 0.24705882, 0.08235294, 0.00784314,
       0.00392157, 0.00392157, 0.02745098, 0.13333334, 0.3137255 ,
       0.52156866, 0.6784314 , 0.7137255 , 0.6039216 , 0.41960785,
       0.23529412, 0.09019608, 0.01176471, 0.01176471, 0.08627451,
       0.22745098, 0.38431373, 0.50980395, 0.53333336, 0.4509804 ,
       0.30980393, 0.1764706 , 0.07058824, 0.01176471, 0.00784314,
       0.05098039, 0.13725491, 0.23529412, 0.30980393, 0.32156864,
       0.27058825, 0.19215687, 0.11372549, 0.05098039, 0.01568628,
       0.00392157, 0.02352941, 0.0627451 , 0.10588235, 0.13333334,
       0.13725491, 0.11764706, 0.08235294, 0.05098039, 0.02352941,
       0.00392157, 0.01176471, 0.01960784, 0.02352941, 0.02352941,
       0.01568628, 0.00784314, 0.00392157, 0.00392157, 0.00392157,
       0.00392157, 0.00392157, 0.00392157, 0.00392157, 0.00392157,
       0.00392157, 0.00392157, 0.00392157, 0.00392157, 0.00392157,
       0.00392157, 0.00392157])

hm = torch.zeros((1, 1, 64, 64))
hm[:, :, x, y] = v
pts = spatial_soft_argmax2d(hm, normalized_coordinates=False)
print(pts)
```

This results in:

```python
tensor([[[31.4091, 31.5107]]])
```

### Expected behavior

Given that `spatial_soft_argmax2d (..., normalized_coordinates=False)` should "_return the coordinates in the range of the input shape_", I would expect the resulting coordinate to be closer to `[-18, 31]`, better matching the blob's observed location in the heatmap image, unless the repro above represents a misuse of the API.

### Environment

```shell
kornia 0.7.0
pytorch 2.0.1
cuda 11.6.2
```

**Comment by edgarriba** [comment]
@markdjwilliams you need to play with the `temperature` parameter. T=15.0 seems a good value for you dataset. https://colab.research.google.com/drive/1CDLKwH93FTFw4mSgyRFmdIjmhtN4pDXG?usp=sharing

**Comment by johnnv1** [comment]
> @markdjwilliams you need to play with the `temperature` parameter. T=15.0 seems a good value for you dataset. https://colab.research.google.com/drive/1CDLKwH93FTFw4mSgyRFmdIjmhtN4pDXG?usp=sharing

did it work for you @markdjwilliams ?

**Comment by markdjwilliams** [comment]
Thank you, this worked, although I was hoping for a solution which would not require tuning against specific datasets - perhaps this is not straightforward?

**Comment by edgarriba** [comment]
Eventually you could learn this parameter.

**Comment by edgarriba** [comment]
If anyone is interested in this, feel free to re-open the ticket. Thanks!
